Ime lib fixes

w1naenator opened the pull request:
- Updated functions to consistently use the Error enum type for return values.
- Added detailed logging to aid future IME/OSK development and debugging.
- Now use OrbisUserServiceUserId (s32) and OrbisImeKeycodeState in relevant functions and structs.
- Introduced a generic template method to generate full bitmasks for all Orbis flag-style enums, simplifying validation and mask creation.
- Implemented additional parameter validations in sceImeOpen.
- Added missing enums: OrbisDisableDevice, OrbisImeInputMethodState, OrbisImeInitExtKeyboardMode, OrbisImeKeycodeState, and other USB keyboard-related enums.
- Fixed incorrect usage of format specifiers in calls to logging macros (LOG_*).

Commit:
IME fixes

- Moved enums, flags, and structs to ime_common.h to simplify usage with Ime and ImeDialog
- Updated Ime to use an enum as the return type, consistent with ImeDialog
- Removed duplicate definition of OrbisImeKeycode
- Added OrbisImeLanguage as a flags enum
- Added missing options to OrbisImeOption
- Removed OrbisImeDialogOption; OrbisImeOption should be used instead
- Added OrbisImeTextAreaMode
- Updated OrbisImeTextAreaMode
- Fixed OrbisImeEventParam by adding the missing member OrbisImePanelType panel_type
- Updated the sceImeOpen declaration to use extended parameters (not yet implemented)
-Fixed Diablo III (CUSA00434) assertion failure on ImeDialog initialization

w1naenator committed:
Merge branch 'main' into ime-fixes

w1naenator committed:
Merge branch 'shadps4-emu:main' into ime-fixes

Commit:
Ime lib fixes
- Updated functions to consistently use the Error enum type for return values.
- Added detailed logging to aid future IME/OSK development and debugging.
- Now use OrbisUserServiceUserId (s32) and OrbisImeKeycodeState in relevant functions and structs.
- Introduced a generic template method to generate full bitmasks for all Orbis flag-style enums, simplifying validation and mask creation.
- Implemented additional parameter validations in sceImeOpen.
- Added missing enums: OrbisDisableDevice, OrbisImeInputMethodState, OrbisImeInitExtKeyboardMode, OrbisImeKeycodeState, and other USB keyboard-related enums.
- Fixed incorrect usage of format specifiers in calls to logging macros (LOG_*).

Commit:
Merge branch 'ime-fixes' of https://github.com/w1naenator/shadPS4 into ime-fixes

georgemoralis commented on src/core/libraries/ime/ime_common.h:
```diff
@@ -3,15 +3,36 @@
 constexpr u32 ORBIS_IME_MAX_TEXT_LENGTH = 2048;
 constexpr u32 ORBIS_IME_DIALOG_MAX_TEXT_LENGTH = 2048;
 
+using OrbisUserServiceUserId = s32;
```
i prefer to use  Libraries::UserService::OrbisUserServiceUserId rather than redefine it

w1naenator replied:
Thank you. I’ve updated it accordingly.

Commit:
Data Type Fixes

- Replaced the use of the type alias OrbisUserServiceUserId = s32 with Libraries::UserService::OrbisUserServiceUserId directly.

w1naenator committed:
Merge branch 'main' into ime-fixes

Commit:
Fixed IDE warnings
- generate_full_mask now returns const instead of constexpr.
- Added argument list to std::unique_lock<std::mutex> construction for clarity.

Commit:
Merge branch 'ime-fixes' of https://github.com/w1naenator/shadPS4 into ime-fixes

Commit:
Clang fixes

georgemoralis commented on src/core/libraries/ime/ime_common.h:
```diff
@@ -3,15 +3,55 @@
 constexpr u32 ORBIS_IME_DIALOG_MAX_TEXT_LENGTH = 2048;
 
+/*
+template <typename E>
```
this needs to be here?

w1naenator replied:
No, it was only for me while I was testing the new version.

Commit:
Removed unneccessary comment

w1naenator committed:
Merge branch 'main' into ime-fixes

StevenMiller123 commented on src/core/libraries/ime/ime.cpp:
```diff
@@ -268,9 +304,12 @@ int PS4_SYSV_ABI sceImeKeyboardGetResourceId() {
 }
 
-Error PS4_SYSV_ABI sceImeKeyboardOpen(s32 userId, const OrbisImeKeyboardParam* param) {
+Error PS4_SYSV_ABI sceImeKeyboardOpen(Libraries::UserService::OrbisUserServiceUserId userId,
+                                      const OrbisImeKeyboardParam* param) {
     LOG_INFO(Lib_Ime, "called");
 
+    LOG_INFO(Lib_Ime, "kValidImeDialogExtOptionMask=0x{:X}", kValidImeDialogExtOptionMask);
+
     if (!param) {
         return Error::INVALID_ADDRESS;
     }
```
I'm not fond of adding more log spam to sceImeKeyboardOpen when it is still bugged, and is spammed by Unity games.
If I remember correctly, these games hit the !param->arg error return, but are programmed to repeatedly call sceImeKeyboardOpen until it returns success.

w1naenator replied:
Can you please provide the name of the game that generates this log spam?

StevenMiller123 replied:
Everything made in Unity spams this, presumably implying the offending code is part of the engine itself rather than the games.
Some examples include Cooking Mama: Cookstar, DIGIMON SURVIVE, Digimon World: Next Order, KINGDOM HEARTS Melody of Memory, NEO : The World Ends with You, Potion Permit, PowerWash Simulator, and Untitled Goose Game.
[CUSA25355.log](https://github.com/user-attachments/files/21322149/CUSA25355.log)
[CUSA26376.log](https://github.com/user-attachments/files/21322150/CUSA26376.log)
[CUSA29983.log](https://github.com/user-attachments/files/21322154/CUSA29983.log)
[CUSA33783.log](https://github.com/user-attachments/files/21322155/CUSA33783.log)
[CUSA05469.log](https://github.com/user-attachments/files/21322156/CUSA05469.log)
[CUSA18223.log](https://github.com/user-attachments/files/21322157/CUSA18223.log)
[CUSA18900.log](https://github.com/user-attachments/files/21322158/CUSA18900.log)
[CUSA23079.log](https://github.com/user-attachments/files/21322159/CUSA23079.log)

w1naenator replied:
Thank you. I will inspect it and try to find a solution.

w1naenator replied:
It seems that param->arg validation is incorrect. There should be a way to check whether arguments are required for the handler, but currently, I'm not performing that check. The Unity engine keeps calling sceImeKeyboardOpen repeatedly until it returns OK.
